### PR TITLE
DRT-1251 Download Manager

### PR DESCRIPTION
**Pull request by jacksonhyde**
- Implements download manager route and components
- Implements redux and redux-saga

### CODE REVIEW

File: react/src/components/Navigation.tsx [hunk 1]
```diff
@@ -30,6 +30,7 @@ export default function Navigation(props: IProps) {
     {label: 'Access requests', link: '/access-requests', roles: ['manage-users']},
     {label: 'Alert notices', link: '/alerts', roles: ['manage-users']},
     {label: 'Drop-in sessions', link: '/drop-in-sessions', roles: ['manage-users']},
+    {label: 'Download Manager', link: '/download', roles: ['manage-users']},
```
Comment (richbirch) [inline react/src/components/Navigation.tsx]: I'm not actually sure which users we're releasing this to. The `manage-users` role means it'll only be available to our own team. We should probably add the RCC roles, which are:
```
rcc:central
rcc:heathrow
rcc:north
rcc:south
```

File: react/src/components/downloadmanager/DownloadManager.tsx [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+  const isRccRegion = (regionName : string) => {
+    return user.roles.includes("rcc:" + regionName.toLowerCase())
+  }
+
+  let interval: {current: ReturnType<typeof setInterval> | null | any} = React.useRef(null);
+
+  React.useCallback(() => {
+    
+  }, [checkDownloadStatus]);
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadManager.tsx]: Do we need this callback if it's not doing anything?

File: react/src/components/downloadmanager/DownloadManager.tsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    end: moment(),
+  });
+  const [breakdown, setBreakdown] = React.useState<string>('passengers-port');
+  const [daily, setDaily] = React.useState<boolean>(false);
+
+  const isRccRegion = (regionName : string) => {
+    return user.roles.includes("rcc:" + regionName.toLowerCase())
+  }
+
+  let interval: {current: ReturnType<typeof setInterval> | null | any} = React.useRef(null);
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadManager.tsx:47]: I think the correct type would be `NodeJS.Timeout | null` as `setInterval` returns a `NodeJS.Timeout`

File: react/src/components/downloadmanager/DownloadManager.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    setModalOpen(false)
+    clearInterval(interval.current);
+  }
+
+  const onBreakdownChange = (event: React.ChangeEvent<HTMLInputElement>, breakdown: string) => {
+    setBreakdown(breakdown);
+  }
+
+  const handlePortChange = (event: SelectChangeEvent<typeof selectedPorts>) => {
+    const {
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadManager.tsx:83]: This is fun! I'd just have written it `const value = event.target.value`

File: react/src/components/downloadmanager/DownloadManager.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+  const handlePortCheckboxGroupChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const {
+      target: { name, checked },
+    } = event;
+    let region = userPortsByRegion.filter(region => region.name === name)[0]
+    if (checked) {
+      // add all in region
+      let newSelection = [...selectedPorts, ...region.ports]
+      let deduped = newSelection.filter((element, index) => {
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadManager.tsx]: These `let`s could all be `const` as far as I can tell - preferring immutability etc

File: react/src/components/downloadmanager/DownloadManager.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        return newSelection.indexOf(element) === index;
+      });
+      setSelectedPorts(deduped)
+    } else {
+      // remove any selected from region
+      setSelectedPorts(selectedPorts.filter(port => !region.ports.includes(port)))
+    }
+  }
+
+  const disablePassengerBreakdown = (): boolean =>  {
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadManager.tsx:114]: This is cool  - I actually didn't know how to write the return type on a parameterless function. Thanks!

File: react/src/components/downloadmanager/DownloadPorts.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                  <FormControlLabel control={<Checkbox name={region.name} onChange={handlePortCheckboxGroupChange} inputProps={{ 'aria-label': 'controlled' }} />} label={region.name} />
+                  <Chip sx={{marginLeft: '1em'}} label={`${region.ports.filter(port => selectedPorts.includes(port)).length} selected`} />
+
+              </AccordionSummary>
+              <AccordionDetails>
+                <Grid container spacing={4}>
+                  <Grid item>
+                    <List sx={{paddingRight: 1, borderRight: region.ports.length > 4 ? '1px solid #eee': 'none'}}>
+                      { region.ports.slice(0,4).map((airport: string) => {
+                          return <ListItem key={airport}><FormControlLabel key={airport} control={<Checkbox name={airport} checked={selectedPorts.includes(airport)} onChange={handlePortCheckboxChange} inputProps={{ 'aria-label': 'controlled' }} />} label={airport} /></ListItem>
```
Comment (richbirch) [inline react/src/components/downloadmanager/DownloadPorts.tsx]: There's scope to move the map contents into a function to reduce the repetition across the 4 rows. No biggie - up to you..

File: react/src/components/downloadmanager/downloadManagerSagas.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    startDate,
+    endDate
+  };
+};
+
+function* handleRequestDownload(action: RequestDownloadActionType) {
+  try {
+    yield put(setStatus('requested'))
+
+    let payload : RequestDownloadPayload = {
```
Comment (richbirch) [inline react/src/components/downloadmanager/downloadManagerSagas.ts]: Nice - the saga magic :)
I think all the `let`s here can also be `const`s

File: react/src/services/ApiClient.ts [hunk 9]
```diff
@@ -30,6 +30,8 @@ export default class ApiClient implements IApiClient {
   public static featureGuidesUpdatePublishedEndpoint = "/api/feature-guides/update-published"
 
   public static exportRegionEndpoint = "/api/export-region"
+  public static exportEndpoint = "/api/export"
+  public static exportStatusEndpoint = "/api/export"
```
Comment (richbirch) [inline react/src/services/ApiClient.ts]: Guess we only need one of these as they're the same value?

**Review by richbirch — APPROVED** [review]
Nice work mate. I just had a few suggestions for tidying, and we'll need to add more roles to the navigation item so non-super users can see it, but otherwise looks good to go!

File: react/src/components/Navigation.tsx (hunk 1 above)
Comment (jacksonhyde) [reply react/src/components/Navigation.tsx]: Ah hah that makes sense!

File: react/src/components/downloadmanager/DownloadManager.tsx (hunk 2 above)
Comment (jacksonhyde) [reply react/src/components/downloadmanager/DownloadManager.tsx]: No definitely don't need - I thought we might need to use it for the status timeout but it wasn't needed in the end, must have forgotten to remove

File: react/src/components/downloadmanager/DownloadManager.tsx (hunk 3 above)
Comment (jacksonhyde) [reply react/src/components/downloadmanager/DownloadManager.tsx:47]: That's what I thought, but the linter throws errors on the clearInterval call, after much faffing about this was the only thing that seemed to build.

Comment (richbirch) [reply react/src/components/downloadmanager/DownloadManager.tsx:47]: bloody linters hey!

File: react/src/components/downloadmanager/DownloadManager.tsx (hunk 4 above)
Comment (jacksonhyde) [reply react/src/components/downloadmanager/DownloadManager.tsx:83]: It is fun isn't it! It's good when you need to declare multiple vars from events (line 102 is a good example). Admittedly when you're just grabbing the value it feels a little OTT but I thought I'd keep it consistent across the handler actions

File: react/src/components/downloadmanager/DownloadManager.tsx (hunk 5 above)
Comment (jacksonhyde) [reply react/src/components/downloadmanager/DownloadManager.tsx]: ah yes ofc, not sure how I let that happen ;)

Comment (richbirch) [reply react/src/components/downloadmanager/DownloadManager.tsx]: very drole!

File: react/src/components/downloadmanager/DownloadPorts.tsx (hunk 7 above)
Comment (jacksonhyde) [reply react/src/components/downloadmanager/DownloadPorts.tsx]: Good shout! Should dry things out a bit

File: react/src/services/ApiClient.ts (hunk 9 above)
Comment (jacksonhyde) [reply react/src/services/ApiClient.ts]: yes, not quite sure why I doubled up!

File: react/src/components/downloadmanager/DownloadPorts.tsx (hunk 7 above)
Comment (richbirch) [reply react/src/components/downloadmanager/DownloadPorts.tsx]: pun city :rofl:

**Review by richbirch — APPROVED** [review]
Smashing!

**Pull request merged**